### PR TITLE
TRA-4075 integrate kratos user management

### PR DESCRIPTION
**Pull request by RamiBerm**

### CODE REVIEW

**Review by mertyildiran — DISMISSED** [review]
Why that diff on `package-lock.json`? `npm install` removes 20k from that lock file.

Also installed and run with `mizu install && mizu view`;

- In case of incorrect password, spinner gets stuck (though I couldn't build the image without changing some code, maybe I did something wrong)

Screenshot from 2022-01-03 22-56-23
Screenshot from 2022-01-03 22-59-34

- Got a `500` on correct form submission, I don't know where exactly;

Screenshot from 2022-01-03 22-56-57

```
[GIN] 2022/01/03 - 19:56:37 | 200 |       33.45µs |      172.17.0.1 | GET      "/echo"
[2022-01-03T19:56:42.643+0000] ERROR ▶ 400 Bad Request ▶ [1 install_controller.go:21 Setup]
[GIN] 2022/01/03 - 19:56:42 | 500 |  276.646935ms |       127.0.0.1 | POST     "/install/"
[2022-01-03T19:56:43.874+0000] ERROR ▶ 400 Bad Request ▶ [1 install_controller.go:21 Setup]
[GIN] 2022/01/03 - 19:56:43 | 500 |  284.884547ms |       127.0.0.1 | POST     "/install/"
[GIN] 2022/01/03 - 19:56:45 | 304 |      81.461µs |       127.0.0.1 | GET      "/manifest.json"
[GIN] 2022/01/03 - 19:56:45 | 304 |        65.9µs |       127.0.0.1 | GET      "/static/css/main.81cf7e04.chunk.css"
[GIN] 2022/01/03 - 19:56:45 | 304 |      53.901µs |       127.0.0.1 | GET      "/static/css/2.dd64bdd0.chunk.css"
[GIN] 2022/01/03 - 19:56:45 | 200 |     307.552µs |       127.0.0.1 | GET      "/static/js/main.582c65fe.chunk.js.map"
[GIN] 2022/01/03 - 19:56:45 | 200 |    8.839276ms |       127.0.0.1 | GET      "/static/js/2.86eb4e6f.chunk.js.map"
[GIN] 2022/01/03 - 19:56:45 | 200 |     432.323µs |       127.0.0.1 | GET      "/static/css/main.81cf7e04.chunk.css.map"
[GIN] 2022/01/03 - 19:56:45 | 200 |     331.882µs |       127.0.0.1 | GET      "/static/css/2.dd64bdd0.chunk.css.map"
[GIN] 2022/01/03 - 19:56:47 | 200 |       29.13µs |      172.17.0.1 | GET      "/echo"
[2022-01-03T19:56:47.431+0000] ERROR ▶ 400 Bad Request ▶ [1 install_controller.go:21 Setup]
[GIN] 2022/01/03 - 19:56:47 | 500 |  336.128382ms |       127.0.0.1 | POST     "/install/"
[GIN] 2022/01/03 - 19:56:57 | 200 |       27.17µs |      172.17.0.1 | GET      "/echo"
```

- Refreshing the page, always falls into "register admin account" state. (maybe because of previous issues)

- Tapper pods are not deployed. Maybe they will be deployed in a later stage?

File: ui/src/components/LoadingOverlay.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+interface LoadingOverlayProps {
+    delay?: number
+}
+
+const LoadingOverlay: React.FC<LoadingOverlayProps> = ({delay}) => {
+
+    const [isVisible, setIsVisible] = useState(false);
+
+    //@ts-ignore
```
[the remark this answers is not in the document]
Comment (RamiBerm) [reply ui/src/components/LoadingOverlay.tsx]: simplified the useEffect, removed isRelevant

File: ui/src/components/LoadingOverlay.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    useEffect(() => {
+        let isRelevant = true;
+
+        setTimeout(() => {
+            if (isRelevant)
+                setIsVisible(true);
+        }, delay ?? SpinnerShowDelayMs);
+
+        return () => isRelevant = false;
+    }, []);
```
[the remark this answers is not in the document]
Comment (RamiBerm) [reply ui/src/components/LoadingOverlay.tsx:18]: removed delay depedency

File: ui/src/helpers/api.js [hunk 3]
```diff
@@ -79,4 +73,71 @@ export default class Api {
...
+        const response = await this.client.get("/install/isNeeded");
+        return response.data;
+    }
+
+    isAuthenticationNeeded = async () => {
+        try {
+            await this.client.get("/status/tap");
+            return false;
+        } catch (e) {
+            if (e.response.status == 401) {
```
[the remark this answers is not in the document]
Comment (RamiBerm) [reply ui/src/helpers/api.js:89]: fixed

File: agent/kratos/kratos.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      lifespan: 10m
+      ui_url: http://127.0.0.1:4455/registration
+      after:
+        password:
+          hooks:
+            -
+              hook: session
+
+log:
+  level: debug
```
Comment (IgorGov) [inline agent/kratos/kratos.yml:58]: default should be info?

File: agent/main.go [hunk 5]
```diff
@@ -268,6 +269,8 @@ func hostApi(socketHarOutputChannel chan<- *tapApi.OutputChannelItem) {
 	routes.EntriesRoutes(app)
 	routes.MetadataRoutes(app)
 	routes.StatusRoutes(app)
+	routes.InstallRoutes(app)
```
Comment (IgorGov) [inline agent/main.go:272]: are these routes for "standalone mode" only?

File: agent/pkg/api/socket_routes.go [hunk 6]
```diff
@@ -44,6 +44,7 @@ func init() {
 	connectedWebsockets = make(map[int]*SocketConnection, 0)
 }
 
+// TODO: add m2m authentication to these routes
```
Comment (IgorGov) [inline agent/pkg/api/socket_routes.go:47]: do we have a jira ticket not to forget to do it?
In "standalone mode" - we should secure the websockets and all api server endpoints

Comment (RamiBerm) [reply agent/pkg/api/socket_routes.go:47]: Created, TRA-4134

File: agent/main.go (hunk 5 above)
Comment (RamiBerm) [reply agent/main.go:272]: They're only relevant there, yes.
We could make them conditional but there's no real reason to, the only endpoint there does nothing when not in standalone mode

File: agent/pkg/controllers/install_controller.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import (
+	"mizuserver/pkg/providers"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+)
+
+func IsSetupNecessary(c *gin.Context) {
+	if IsInstallNeeded, err := providers.IsInstallNeeded(); err != nil {
+		c.AbortWithStatusJSON(500, gin.H{"error": "internal error occured while checking if install is needed"})
```
[the remark this answers is not in the document]
Comment (RamiBerm) [reply agent/pkg/controllers/install_controller.go:12]: Definitely needs printing

File: agent/kratos/kratos.yml (hunk 4 above)
Comment (RamiBerm) [reply agent/kratos/kratos.yml]: updated

File: agent/kratos/start.sh [hunk 8]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+kratos migrate sql sqlite:///app/data/kratos.sqlite?_fk=true --yes # this initializes the db
+kratos serve -c /etc/config/kratos/kratos.yml --dev --watch-courier # start kratos
```
Comment (IgorGov) [inline agent/kratos/start.sh:4]: --dev is for development?

File: agent/main.go (hunk 5 above)
Comment (IgorGov) [reply agent/main.go:272]: lets hide same as here
https://github.com/up9inc/mizu/blob/bf686892123755b16eea4eece8bc393a2a76c080/agent/main.go#L263

File: agent/kratos/start.sh (hunk 8 above)
Comment (RamiBerm) [reply agent/kratos/start.sh]: good catch, not needed anymore 
removed it

File: cli/resources/createResources.go [hunk 9]
```diff
@@ -65,6 +66,12 @@ func CreateTapMizuResources(ctx context.Context, kubernetesProvider *kubernetes.
 }
 
 func CreateInstallMizuResources(ctx context.Context, kubernetesProvider *kubernetes.Provider, serializedValidationRules string, serializedContract string, serializedMizuConfig string, isNsRestrictedMode bool, mizuResourcesNamespace string, agentImage string, syncEntriesConfig *shared.SyncEntriesConfig, maxEntriesDBSizeBytes int64, apiServerResources shared.Resources, imagePullPolicy core.PullPolicy, logLevel logging.Level, noPersistentVolumeClaim bool) error {
+	if !isNsRestrictedMode {
```
Comment (IgorGov) [inline cli/resources/createResources.go:69]: we should not support his mode in standalone mode

**Review by IgorGov — APPROVED** [review]

**Review by IgorGov — APPROVED** [review]

**Review by IgorGov — APPROVED** [review]

**Pull request merged**